### ISSUE TITLE
RequireThisCheck giving multiple violation for classes nested in lambdas.

### ISSUE THREAD

**Issue by Vyom-Yadav** [checkstyle/checkstyle#11822]
I have read check documentation: https://checkstyle.sourceforge.io/config_coding.html#RequireThis
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac Test.java
```
```bash
/var/tmp $ cat config.xml
```
```XML
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.2//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_2.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="RequireThisCheck">
      <property name="validateOnlyOverlapping" value="false" />
    </module>
  </module>
</module>
```
```bash
/var/tmp $ cat Test.java
```
```java
import java.util.function.Consumer;

public interface Test {

    private static void method(Consumer<String> consumer) {
        consumer.accept("foo");
    }

    private void testCtorNestedInLambda() {
        method(s -> {
            class InnerSubscriber implements Test {
                int index;

                public InnerSubscriber(int index) {
                    index = index; // 2 violations, expected 1
                }
            }
        });
    }
}
```
```bash
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.3.1-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:15:21: Reference to instance variable 'index' needs "this.". [RequireThis]
[ERROR] /var/tmp/Test.java:15:29: Reference to instance variable 'index' needs "this.". [RequireThis]
Audit done.
Checkstyle ends with 2 errors.
```
It is giving violation for both sides, the check is trying to imply `this.index = this.index` which isn't correct.

**Comment by Kevin222004** [comment]
I was trying to figure out this issue for the past 2 days still. I am not able to figure out the whole code. I have got only basic idea after giving lots of time. is their anything special (any algorithm or any concept) in this check which I should learn to make it easy

**Comment by rnveach** [comment]
2 violations started appearing in 8.27 , but only 1 appeared in 8.26 . Maybe we changed something and broke something?

````
$ cat TestClass.java
import java.util.function.Consumer;

public interface Test {

    private static void method(Consumer<String> consumer) {
        consumer.accept("foo");
    }

    private void testCtorNestedInLambda() {
        method(s -> {
            class InnerSubscriber implements Test {
                int index;

                public InnerSubscriber(int index) {
                    index = index; // 2 violations, expected 1
                }
            }
        });
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>

    <module name="TreeWalker">

    <module name="RequireThisCheck">
      <property name="validateOnlyOverlapping" value="false" />
    </module>

    </module>
</module>

$ java -jar checkstyle-8.26-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[WARN] TestClass.java:15:21: Reference to instance variable 'index' needs "this.". [RequireThis]
Audit done.

$ java -jar checkstyle-8.27-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[WARN] TestClass.java:15:21: Reference to instance variable 'index' needs "this.". [RequireThis]
[WARN] TestClass.java:15:29: Reference to instance variable 'index' needs "this.". [RequireThis]
Audit done.
````

**Comment by AayushSaini101** [comment]
@rnveach @romani I am on it 

I have a doubt is the correct violation:   this.booleanField = booleanField; // violation '.* variable 'booleanField' needs "this.".'

Config: 

```
RequireThis
checkFields = (default)true
checkMethods = (default)true
validateOnlyOverlapping = false
```


FileName: InputRequireThisValidateOnlyOverlappingFalse.java 
 ```
 void foo24() {
        String field1 = "Hello";
        field1 = "Java"; // No violation. Local var allowed
        this.booleanField = true;
        this.booleanField = booleanField; // violation '.* variable 'booleanField' needs "this.".'
    }
```
I think this is not a violation,

**Comment by romani** [comment]
@AayushSaini101,
please make CLI output as we use in issue template, your example `booleanField` is not defined as variable, so it is field.
and  
`this.booleanField = booleanField; // violation '.* variable 'booleanField' needs "this.".'`
should be
`this.booleanField = this.booleanField;`
as Check demands `this.` for all fields at any place.

**Comment by vivek-0509** [comment]
i am on it

**Comment by romani** [comment]
Fix is merged
